lib/wibox/layout/grid.lua

```diff
@@ -881,11 +1332,171 @@ function grid:layout(context, width, height)
         -- Place the widget if it fits in the area
         if x + w <= width and y + h <= height then
             table.insert(result, base.place_widget_at(v.widget, x, y, w, h))
+            table.insert(areas, {
+                x      = x - hspacing,
+                y      = y - vspacing,
+                width  = w + col_spacing,
+                height = h + row_spacing,
+            })
         end
     end
-    return result
+
+    -- Sometime, the `:fit()` size and `:layout()` size are different, thus it's
+    -- important to say where the widget actually ends.
+    areas.end_x = cumul_width[#cumul_width] - hspacing
+    areas.end_y = cumul_height[#cumul_height] - vspacing
+    areas.column_count = #cols_size
+    areas.row_count = #rows_size
+    areas.cols = cumul_width
+    areas.rows = cumul_height
+
+    return result, areas
+end
+
+-- Layout a grid layout.
+-- @param context The context in which we are drawn.
+-- @param width The available width.
+-- @param height The available height.
+function grid:layout(context, width, height)
+    local l, areas = layout_common(
+        self,
+        context,
+        width,
+        height,
+        self._private.horizontal_homogeneous,
+        self._private.vertical_homogeneous
+    )
+
+    self._private.area_cache[width*1.5+height*15] = areas
```

Comment:
Can we have an utility function 

```lua
local function get_area_cache_hash(width, height)
   return width*1.5+height*15
end
```

So that we are sure to use the same key everywhere we want to compute this hash.

Reply:
done